xtc/src/history.rs

```diff
@@ -71,12 +75,72 @@ impl HistoryBuffer {
     pub fn len(&self) -> u64 {
         self.history.len()
     }
+
+    pub fn history(&self) -> &History {
+        &self.history
+    }
+}
+
+//////////////////// BEGIN OF ERC-20 ///////////////////////
+
+#[update(name = "getTransaction")]
+pub async fn get_transaction_erc20(index: Nat) -> TxRecord {
+    TryInto::<TxRecord>::try_into(
+        get_context()
+            .get::<HistoryBuffer>()
+            .history
+            .get_transaction(convert_nat_to_u64(index.clone()).unwrap())
+            .await
+            .expect("cannot find the transaction in the history"),
+    )
+    .expect("unable to convert the retrieved transaction into TxRecord")
+    .setIndex(index)
 }
 
+// TODO: ticking time bomb, we need to integrate the history service, as we
+// are always using the internal buffer here
+#[update(name = "getTransactions")]
+pub async fn get_transactions(start: Nat, limit: Nat) -> Vec<TxRecord> {
+    let MAX_LIMIT = Nat::from(100);
+
+    if limit > MAX_LIMIT {
+        ic_cdk::api::trap(&format!("limit cannot be greater than {}", MAX_LIMIT))
+    }
+
+    let start_usize = convert_nat_to_u64(start).unwrap() as usize;
+    let limit_usize = convert_nat_to_u64(limit).unwrap() as usize;
+    let events = get_context()
+        .get::<HistoryBuffer>()
+        .history()
+        .get_history_data()
+        .get_events();
+    let history_usize = events.len();
+
+    if (start_usize >= history_usize) {
+        ic_cdk::api::trap(&format!(
+            "start cannot be greater than the history size of {}",
+            history_usize
+        ))
+    }
+
+    (&events[start_usize..min(history_usize, start_usize + limit_usize)])
+        .into_iter()
+        .enumerate()
+        .filter_map(|tx_pair| match (tx_pair.1.clone()).try_into().ok() {
+            Some(tx) => Some((tx_pair.0, tx)),
+            _ => None,
+        })
```

Comment:
The filter is not needed if you use `From` as i suggested above. Then you can move the `enumerate` after the map as well.